posts/react-compound-components/2021-5-25-React-compound-compnents.md

```diff
@@ -0,0 +1,250 @@
+
+## What are compound components?
+
+The compound components are a set of two or more components that work together to accomplish a specific task. The set of components will share an implicit state to communicate between them.
+
+> Think of compound components like the select and option elements in HTML. Apart they don’t do too much, but together they allow you to create the complete experience. — Kent C. Dodds
+
+```jsx
+<select>
+ <option>Option1</option>
+ <option>Option2</option>
+ <option>Option3</option>
+ <option>Option4</option>
+</select>
+```
+
+In the above example, when you click on an option in the select component, select knows which option you clicked. The select and the option share the state between them and update the selected option state on their own, we don’t need to explicitly configure them.
+
+## Compound Components in Action
+
+In this post, we’ll be building an Accordion component using compound components.
+
+You can check out the [Code sandbox link](https://codesandbox.io/s/interesting-wildflower-wj3iy) for the final demo of the accordion component.
+
+The accordion component will have four components.
+
+1. **Accordion** -  The outer wrapper component of the Accordion component. This is the root component of the Accordion component.
+2. **AccordionItem** - The component that allows us to define each accordion item. Each AccordionItem will have its AccordionButton and AccordionPanel components.
+3. **AccordionButton** - The header for the Accordion component.On clicking the accordion button will open the corresponding accordion panel.
+4. **AccordionPanel** - The panel for the accordion. This will hold the content of each accordion item.
+
+We are going to create the above mentioned components one by one and also let's see how we can create the link between them. 
+Let's start with Accordion component. Accordion component will wrap all other necessary components and will maintain the state that is to be shared among all the other components
+
+```tsx
+const Accordion: React.FC<{
+  children: ReactNode | ReactNode[];
+  className?: string
+}> = ({ children, className }) => {
+  const [activeItem, setActiveItem] = useState("");
+  // function to update the active item
+   const changeActiveItem = useCallback(
+    (value) => {
+      if (activeItem !== value) setActiveItem(value);
+    },
+    [setActiveItem, activeItem]
+  );
+
+  return <div className={className}>{children}</div>
+}
+``` 
+
+we had created the accordion component, now we need to pass or share the state and the function to update the state to its children. To achieve this, we are going to use the **React Context**. If you are not familiar with React context, please refer  [https://reactjs.org/docs/context.html](https://reactjs.org/docs/context.html). Iam not gonna explain about react context here, but i will give you a hint about react context.
```

Comment:
We have created the root Accordion component. Now we need to pass down or share the state and the update state function with its children.

Reply:
Please remove this sentence
"Iam not gonna explain about react context here, but i will give you a hint about react context."